File: frontend/src/api/compare.ts
import http from './http'

export type PeriodType = 'annual' | 'quarterly'
export type StatementType = 'balance' | 'income' | 'cashflow'

export interface ComparePeriod {
  year: number
  period_type: PeriodType
  quarter: number | null
  has_balance: boolean
  has_income: boolean
  has_cashflow: boolean
}

export interface ComparePeriodMeta {
  year: number
  period_type: PeriodType
  quarter: number | null
  label: string
  statement_id: number | null
}

export interface CompareFieldRow {
  key: string
  label: string
  values: Array<number | null>
  deltas: Array<number | null>
  delta_pcts: Array<number | null>
  structure_pcts: Array<number | null>
}

export interface CompareGroup {
  key: string
  label: string
  rows: CompareFieldRow[]
}

export interface CompareMatrix {
  company_id: number
  statement_type: StatementType
  period_type: PeriodType
  base_field: string | null
  base_label: string | null
  periods: ComparePeriodMeta[]
  groups: CompareGroup[]
}

export async function fetchComparePeriods(
  companyId: number
): Promise<ComparePeriod[]> {
  const { data } = await http.get<ComparePeriod[]>(
    `/api/companies/${companyId}/compare-periods`
  )
  return data
}

export async function fetchCompare(
  companyId: number,
  params: {
    statement_type: StatementType
    period_type?: PeriodType
    years?: number[]
  }
): Promise<CompareMatrix> {
  const { data } = await http.get<CompareMatrix>(
    `/api/companies/${companyId}/compare`,
    {
      params: {
        statement_type: params.statement_type,
        period_type: params.period_type ?? 'annual',
        years: params.years?.length ? params.years.join(',') : undefined,
      },
    }
  )
  return data
}

/** 金额千分位；null → — */
export function formatAmount(value: number | null | undefined): string {
  if (value === null || value === undefined || Number.isNaN(value)) return '—'
  return value.toLocaleString('zh-CN', {
    maximumFractionDigits: 2,
    minimumFractionDigits: 0,
  })
}

/** 变动率 / 结构占比 → 百分比文案（输入为 0–1 比例） */
export function formatPct(value: number | null | undefined): string {
  if (value === null || value === undefined || Number.isNaN(value)) return '—'
  const pct = value * 100
  const sign = pct > 0 ? '+' : ''
  return `${sign}${pct.toFixed(1)}%`
}

File: backend/app/schemas/compare.py
"""多期科目对比 API 响应模型。"""
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


class ComparePeriodMeta(BaseModel):
    year: int
    period_type: Literal["annual", "quarterly"]
    quarter: int | None = None
    label: str
    statement_id: int | None = None


class CompareFieldRow(BaseModel):
    key: str
    label: str
    values: list[float | None]
    deltas: list[float | None]
    delta_pcts: list[float | None]
    structure_pcts: list[float | None]


class CompareGroup(BaseModel):
    key: str
    label: str
    rows: list[CompareFieldRow] = Field(default_factory=list)


class CompareMatrix(BaseModel):
    company_id: int
    statement_type: Literal["balance", "income", "cashflow"]
    period_type: Literal["annual", "quarterly"]
    base_field: str | None = None
    base_label: str | None = None
    periods: list[ComparePeriodMeta]
    groups: list[CompareGroup]


# 期间列表复用比率侧结构（字段一致）
class ComparePeriod(BaseModel):
    year: int
    period_type: Literal["annual", "quarterly"]
    quarter: int | None = None
    has_balance: bool = False
    has_income: bool = False
    has_cashflow: bool = False
